portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,68 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*; 
+import com.google.gson.Gson;
+import com.google.sps.data.Comment;
+
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
+    Entity commentEntity = new Entity("comment");
+
+    String comment = request.getParameter("comment");
+    String name = request.getParameter("name");
+    Date date = new Date();
+
+    commentEntity.setProperty("date", date);
+    commentEntity.setProperty("name", name);
+    commentEntity.setProperty("comment", comment);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+    
+    response.sendRedirect("/index.html");
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Yuyan!</h1>");
+
+    Query query = new Query("comment").addSort("date", SortDirection.DESCENDING);
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery result = datastore.prepare(query);
+
+    List<Comment> commentList = new ArrayList<>();
+
+    for (Entity entity : result.asIterable()) {
+      String name = (String) entity.getProperty("name");
+      String comment = (String) entity.getProperty("comment");
+      Date date = (Date) entity.getProperty("date");
+
+      Comment commentObj = new Comment(name, comment, date);
+      commentList.add(commentObj);
+    }
+
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(commentList));
```

Comment:
I'm guessing that you are making several temporary strings in this code that didn't actually have to be made. Does Gson operate in a streaming way? You should be able to write the JSON data directory to the request in some fashion. Consider this for a subsequent change.

Reply:
Could you explain what you mean by JSON data directory and the temporary strings? JSON is something new to me, so I'm probably not understanding this fully. Could we possibly chat about this a little more at a later time? Thanks.

Reply:
Sorry. First, I'm not asking you to change anything. I'm simply observing that I think that the line

```
response.getWriter().println(gson.toJson(commentList));
```

is conceivably not as efficient as it could be because `gson.toJson(commentList)` is [making a String](https://www.javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/Gson.html)

Using a variant that writes to the `getWriter()` stream would not have to make a temporary copy of the serialized JSON data in memory. For large JSON, this has some performance benefits.